[dagster-components] Introduce `load_defs` entrypoint, which requires module instead of path

benpankow opened the pull request:
## Summary

In order to facilitate more module-oriented functionality (#28133), we introduce a new `load_defs` entrypoint to replace `build_component_defs`, which takes a `ModuleType` as input.

```python
from pathlib import Path

from dagster_components import build_defs
import my_components_proj.defs

defs = load_defs(defs_root=my_components_proj.defs)
```


## Test Plan

Existing unit tests.

benpankow commented:
* **#28264** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28264?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28264?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benpankow commented on examples/docs_beta_snippets/docs_beta_snippets/guides/components/index/5-definitions.py:
```diff
@@ -1,5 +1,6 @@
 from pathlib import Path
 
 from dagster_components import build_component_defs
+from jaffle_platform import defs as component_defs
```
this pattern is unfortunate

schrockn commented:
I'm a bit confused. Seems strictly better to do the module reference. @smackesey can you provide a concrete example where a string is preferable?

smackesey commented:
>I'm a bit confused. Seems strictly better to do the module reference. @smackesey can you provide a concrete example where a string is preferable?

Depends on the context in which this function will be called. If it is only called like in the snippet in OP, no difference. A string defers the import and gives the function control over any errors that occur during import, module caching, etc. This control can be useful depending on the context in which the function is called. But like I said, I'm fuzzy on where and how this function will be used in the future. Weakly held opinion so I have no objections if you think a module ref is better.

schrockn commented on python_modules/libraries/dagster-components/dagster_components/core/component_defs_builder.py:
```diff
@@ -97,13 +107,24 @@ def build_component_defs(
     """
     from dagster._core.definitions.definitions_class import Definitions
 
+    # For backcompat, allow a path to be passed, even though the signature doesn't support it
+    if isinstance(components_root, Path):
+        warnings.warn(
+            "Passing a path to build_component_defs is deprecated. Pass a module instead.",
+            DeprecationWarning,
+        )
+        components_root = f"{Path(components_root).parent.name}.{Path(components_root).name}"
```
why are we deprecating this?

schrockn reviewed (dismissed):
why are we changing top-level signature to use `str` instead of `Path`?

benpankow commented:
Per offline conversation, renamed to `build_defs`, and reverted to taking a `ModuleType`. Retained `build_component_defs` with a deprecated decorator, for backcompat.

schrockn commented on docs/docs/guides/labs/components/building-pipelines-with-components/pyproject-toml.md:
```diff
@@ -29,7 +29,7 @@ To understand the structure, let's look at the content of `jaffle_platform/defin
 <CodeExample path="docs_beta_snippets/docs_beta_snippets/guides/components/index/5-definitions.py" language="Python" title="jaffle-platform/jaffle_platform/definitions.py" />
 
-This call to `build_component_defs` will:
+This call to `build_defs` will:
```
How about we call this `autoload_defs` to disambiguate from `build_defs`? They are pretty different operations. I think `autoload` will also indicate just how heavy it is

schrockn replied:
`load_defs`

schrockn reviewed (changes requested):
naming fun. 

`autoload_defs`? `load_defs`?

I'm just imagining the language we will use to refer to the process of loading all the definitions via the this process.

"On startup, `dg` discovers and loads all definitions in the `defs` module via an  autoloading process. It recursively finds all modules that contains definition and loads them into memory. Further more it discovers any component declarations and loads the definitions that component generates"

schrockn reviewed (changes requested):
Per discussion let's call this `load_defs`

schrockn commented on docs/docs/guides/labs/components/incrementally-adopting-components/existing-project.md:
```diff
@@ -66,7 +66,7 @@ Next, you'll want to create a directory to contain any new components you add to
 To do so, you'll need to modify your `definitions.py` file, or whichever file contains your top-level `Definitions` object.
 
-You can manually construct a set of definitions for your components using `build_component_defs`, then merge them with your existing definitions using `Definitions.merge`. You point `build_components_defs` at the directory you just created that contains components.
+You can manually construct a set of definitions for your components using `build_defs`, then merge them with your existing definitions using `Definitions.merge`. You point `build_components_defs` at the directory you just created that contains components.
```
replace with `load_defs`

schrockn commented on examples/docs_beta_snippets/docs_beta_snippets_tests/snippet_checks/guides/components/test_components_docs_existing_project/test_existing_code_location.py:
```diff
@@ -120,7 +121,7 @@ def test_components_docs_index(update_snippets: bool) -> None:
                 defs = dg.Definitions.merge(
                     dg.Definitions(assets=all_assets),
-                    dg_components.build_component_defs(Path(__file__).parent / "components"),
+                    dg_components.build_defs(my_existing_project.defs),
```
load_defs

schrockn commented on examples/docs_beta_snippets/docs_beta_snippets_tests/snippet_checks/guides/components/test_components_docs_migrating_definitions/test_migrating_definitions.py:
```diff
@@ -204,8 +205,9 @@ def test_components_docs_migrating_definitions(update_snippets: bool) -> None:
                 from pathlib import Path
 
                 import dagster_components as dg_components
+                import my_existing_project.defs
 
-                defs = dg_components.build_component_defs(Path(__file__).parent / "defs")
+                defs = dg_components.build_defs(my_existing_project.defs)
```
<!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'><i>[Re: lines +128 to +210]</i></p><!--__GRAPHITE_HTML_TAG_END__-->
load_defs
<!--__GRAPHITE_HTML_TAG_START__--><p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/28264?utm_source=unchanged-line-comment">Graphite</a>.</p><!--__GRAPHITE_HTML_TAG_END__-->

schrockn commented on python_modules/libraries/dagster-components/dagster_components_tests/integration_tests/component_loader.py:
```diff
@@ -19,8 +20,9 @@ def load_test_component_defs(
-            path=Path(code_location_dir) / "my_location" / "defs" / Path(src_path).stem,
+        module = importlib.import_module(f"my_location.defs.{Path(src_path).stem}")
+
+        return build_defs_from_component_module(
```
`load_defs`

schrockn commented on python_modules/libraries/dagster-dg/dagster_dg/templates/PROJECT_NAME_PLACEHOLDER/PROJECT_NAME_PLACEHOLDER/definitions.py.jinja:
```diff
@@ -0,0 +1,6 @@
+import {{ project_name }}.defs
+from dagster_components import build_defs
+
+defs = build_defs(defs_root={{ project_name }}.defs)
```
`load_defs`

schrockn reviewed (approved):
Looks good in context. You need to cleanup some final stuff.
